DQM/SiStripMonitorClient/plugins/SiStripBadComponentInfo.cc

```diff
@@ -0,0 +1,232 @@
+      val += range;
+      badStripME_->setBinContent(subdet, component,val);
+    }
+  }
+}
+//
+// -- End Run
+//
+void SiStripBadComponentInfo::dqmEndJob(DQMStore::IBooker &ibooker, DQMStore::IGetter &igetter){
+  LogDebug ("SiStripBadComponentInfo") <<"SiStripBadComponentInfo::dqmEndRun";
+  bookBadComponentHistos(ibooker, igetter);
+  checkBadComponents();  
+  createSummary(badAPVME_); 
+  createSummary(badFiberME_); 
+  createSummary(badStripME_); 
+} 
+//
+// -- Book MEs for SiStrip Dcs Fraction
+//
+void SiStripBadComponentInfo::bookBadComponentHistos(DQMStore::IBooker &ibooker, DQMStore::IGetter &igetter) {
+  if (!bookedStatus_) {
+    std::string strip_dir = "";
+    ibooker.cd();
+    //    SiStripUtility::getTopFolderPath(dqmStore_, "SiStrip", strip_dir); 
+    if (igetter.dirExists("SiStrip")) {
+      ibooker.cd("SiStrip");
+      strip_dir = ibooker.pwd();
+    }
+    std::cout << "SiStripBadComponentInfo::bookBadComponentHistos ==> " << strip_dir << " " << ibooker.pwd() << std::endl;
+    if (!strip_dir.empty()) ibooker.setCurrentFolder(strip_dir+"/EventInfo");
+    else ibooker.setCurrentFolder("SiStrip/EventInfo");
+       
+    ibooker.cd();
+    if (!strip_dir.empty())  ibooker.setCurrentFolder(strip_dir+"/EventInfo/BadComponentContents");
+
+    std::string  hname, htitle;
+    hname  = "BadAPVMap";
+    htitle = "SiStrip Bad APVs";
+    badAPVME_ = ibooker.book2D(hname, htitle, nSubSystem_, 0.5, nSubSystem_+0.5, 10, 0.5, 10.5);
+    badAPVME_->setAxisTitle("Sub Detector Type", 1);
+    badAPVME_->setAxisTitle("Layer/Disc Number", 2);
+
+    hname  = "BadFiberMap";
+    htitle = "SiStrip Bad Fibers";
+    badFiberME_ = ibooker.book2D(hname, htitle, nSubSystem_, 0.5, nSubSystem_+0.5, 10, 0.5, 10.5);
+    badFiberME_->setAxisTitle("Sub Detector Type", 1);
+    badFiberME_->setAxisTitle("Layer/Disc Number", 2);
+
+    hname  = "BadStripMap";
+    htitle = "SiStrip Bad Strips";
+    badStripME_ = ibooker.book2D(hname, htitle, nSubSystem_, 0.5, nSubSystem_+0.5, 10, 0.5, 10.5);
+    badStripME_->setAxisTitle("Sub Detector Type", 1);
+    badStripME_->setAxisTitle("Layer/Disc Number", 2);
+
+    std::vector<std::string> names;
+    names.push_back("TECB");
+    names.push_back("TECF");
+    names.push_back("TIB");
+    names.push_back("TIDB");
+    names.push_back("TIDF");
+    names.push_back("TOB");
+
+    for (unsigned int i=0; i < names.size(); i++) {
+      badAPVME_->setBinLabel(i+1, names[i]);       
+      badFiberME_->setBinLabel(i+1, names[i]);       
+      badStripME_->setBinLabel(i+1, names[i]);       
+    }
+
+    bookedStatus_ = true;
+    ibooker.cd();
+  }
+}
+void SiStripBadComponentInfo::fillBadComponentHistos(int xbin,int component,SiStripQuality::BadComponent& BC){
+
+  if (BC.BadApvs){ 
+    int ntot =  ( (BC.BadApvs>>5)&0x1 )+ ( (BC.BadApvs>>4)&0x1 ) + ( (BC.BadApvs>>3)&0x1 ) + 
+      ( (BC.BadApvs>>2)&0x1 )+ ( (BC.BadApvs>>1)&0x1 ) + ( (BC.BadApvs)&0x1);
```

Comment:
@suchandradutta, it was pointed out to me in another occurrence of the same code, that this line can be heavily simplified using `std::bitset`:

```C++
int ntot = std::bitset<16>(BC.BadApvs&0x3f).count(); 
```